include/deal.II/base/tensor.h

```diff
@@ -1749,12 +1698,30 @@ constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
   typename numbers::NumberTraits<Number>::real_type
   Tensor<rank_, dim, Number>::norm_square() const
 {
-  typename numbers::NumberTraits<Number>::real_type s = internal::NumberType<
-    typename numbers::NumberTraits<Number>::real_type>::value(0.0);
-  for (unsigned int i = 0; i < dim; ++i)
-    s += values[i].norm_square();
+  if constexpr (rank_ == 1)
+    {
+      // For rank-1 tensors, the square of the norm is simply the sum of
+      // squares of the elements:
+      typename numbers::NumberTraits<Number>::real_type s =
+        internal::NumberType<
+          typename numbers::NumberTraits<Number>::real_type>::value(0.0);
+      for (unsigned int i = 0; i < dim; ++i)
+        s += numbers::NumberTraits<Number>::abs_square(values[i]);
```

Comment:
I'm usually preferring to peel off the first loop index because it avoids adding into zero (and compilers would need `-ffast-math` to avoid the instruction); in this case, it wouldn't even introduce more lines of code because we would write
```suggestion
      typename numbers::NumberTraits<Number>::real_type s =
        numbers::NumberTraits<Number>::abs_square(values[0]);
      for (unsigned int i = 1; i < dim; ++i)
        s += numbers::NumberTraits<Number>::abs_square(values[i]);
```
The question that remains for me is if would trigger a warning for `dim = 0`. I do not feel particularly strongly about it, so either one is fine for me.

Reply:
This is a loop with compile-time bounds. It's almost certainly going to be unrolled. Don't you think the compiler can make the optimization itself? I think it's easier to read this way, but I also don't feel strongly about it.

Reply:
I changed my mind and did as you suggested. I handled the `dim==0` case explicitly via an other `if constexpr` at the top.